pandas/core/generic.py

```diff
@@ -6113,20 +6113,24 @@ def fillna(
                 )
 
             elif isinstance(value, (dict, ABCSeries)):
+                new_data = self if inplace else self.copy()
+
                 if axis == 1:
-                    raise NotImplementedError(
-                        "Currently only can fill "
-                        "with dict/Series column "
-                        "by column"
-                    )
+                    for i, item in enumerate(new_data.items()):
+                        label, content = item
+                        new_data.iloc[:, i] = content.fillna(
+                            value, limit=limit, inplace=False, downcast=downcast
+                        )
+                else:
+                    for i, item in enumerate(new_data.items()):
+                        label, content = item
+                        if label not in value:
+                            continue
+                        new_data.iloc[:, i] = content.fillna(
+                            value[label], limit=limit, inplace=False, downcast=downcast
+                        )
 
-                result = self if inplace else self.copy()
-                for k, v in value.items():
-                    if k not in result:
-                        continue
-                    obj = result[k]
-                    obj.fillna(v, limit=limit, inplace=True, downcast=downcast)
-                return result if not inplace else None
+                new_data = new_data.infer_objects()
```

Comment:
i think mypy is complaining because it expects new_data to be a BlockManager, so maybe use a different name for this?